packages/iris-grid/src/IrisGridTableModelTemplate.ts

```diff
@@ -1893,17 +1895,59 @@ class IrisGridTableModelTemplate<
 
   async setValues(edits: readonly EditOperation[] = []): Promise<void> {
     log.debug('setValues(', edits, ')');
-    if (
-      !edits.every(edit =>
-        this.isEditableRange(
-          GridRange.makeCell(edit.column ?? edit.x, edit.row ?? edit.y)
-        )
-      )
-    ) {
+
+    const invalidEdits = edits.filter(edit => {
+      const column = edit.column ?? edit.x;
+      const row = edit.row ?? edit.y;
+      const range = GridRange.makeCell(column, row);
+
+      if (this.isEditableRange(range)) {
+        return false;
+      }
+
+      if (
+        this.inputTable != null &&
+        range.startRow != null &&
+        range.startColumn != null
+      ) {
+        const pendingAreaStart = this.floatingTopRowCount + this.table.size;
+        if (range.startRow >= pendingAreaStart) {
+          const isKeyColumn = this.isKeyColumn(range.startColumn);
+          const hasKeyColumns = this.keyColumnSet.size > 0;
+          return hasKeyColumns && isKeyColumn;
+        }
+      }
+
+      return true;
+    });
+
+    if (invalidEdits.length > 0) {
       throw new Error(`Uneditable ranges ${edits}`);
     }
```

Comment:
Before the change we would throw on the first uneditable range, without needing to iterate through the remaining ranges. Now we're always iterating through every one, even though we throw if any of them are invalid. That might make sense if we included the list of invalid ranges in the thrown error, but we don't - we just list all the ranges.
This logic should just be incorporated into the `isEditableRange` method anyway, and this should just be calling that method (as it was before). Anything else that is checking for if something is an editable range should have the same logic.